File: apps/react-node/server/outbound-http-logging.ts
import { writeLog } from "./observability-logging.js";
import {
  currentInboundMethod,
  currentInboundPath,
  currentRequestId,
  currentSessionId,
} from "./request-context.js";
import { outboundRequestHeaders, resolveOutboundRequestId } from "./request-id.js";

const LOGGER = "http.client";
const RELAY_ORIGIN = "exercises-react-node";

const HEADER_ALLOW = new Set([
  "x-request-id",
  "x-request-origin",
  "content-type",
  "content-length",
  "server",
  "date",
  "location",
]);

function pathWithQuery(url: string): string {
  try {
    const parsed = new URL(url);
    const pathname = parsed.pathname || "/";
    return parsed.search ? `${pathname}${parsed.search}` : pathname;
  } catch {
    return url;
  }
}

function resolveRequestIdForLog(requestId?: string): string {
  const explicit = requestId?.trim();
  if (explicit) {
    return explicit;
  }
  const inbound = currentRequestId();
  if (inbound) {
    return inbound;
  }
  return resolveOutboundRequestId();
}

function parseBody(raw: string | null | undefined): Record<string, unknown> {
  if (!raw?.trim()) {
    return {};
  }
  const text = raw.trim();
  try {
    if (text.startsWith("{")) {
      return JSON.parse(text) as Record<string, unknown>;
    }
    if (text.startsWith("[")) {
      return { _json: JSON.parse(text) as unknown };
    }
  } catch {
    // fall through to raw string
  }
  return { _raw: text };
}

export function resolveResponseError(
  status: number,
  responseBody: string | null | undefined,
): string | null {
  if (status >= 200 && status < 300) {
    return null;
  }
  const parsed = parseBody(responseBody);
  const err = parsed.error;
  if (typeof err === "string" && err.trim()) {
    return err.trim();
  }
  const message = parsed.message;
  if (typeof message === "string" && message.trim()) {
    return message.trim();
  }
  const raw = parsed._raw;
  if (typeof raw === "string" && raw.trim()) {
    return raw.trim();
  }
  if (Object.keys(parsed).length > 0) {
    return JSON.stringify(parsed);
  }
  return `HTTP ${status}`;
}

function filterResponseHeaders(headers: Headers): Record<string, string> {
  const out: Record<string, string> = {};
  headers.forEach((value, name) => {
    if (HEADER_ALLOW.has(name.toLowerCase())) {
      out[name.toLowerCase()] = value;
    }
  });
  return out;
}

function emit(level: "INFO" | "WARN", message: string, fields: Record<string, unknown>): void {
  writeLog(level, message, fields, LOGGER);
  const line = JSON.stringify({ level, message, logger: LOGGER, ...fields });
  if (level === "WARN") {
    console.warn(line);
  } else {
    console.info(line);
  }
}

function baseFields(
  method: string,
  url: string,
  relayTarget: string,
  requestId: string,
  phase: string,
  extra?: Record<string, unknown>,
): Record<string, unknown> {
  const sessionId = currentSessionId();
  const originMethod = currentInboundMethod();
  const originPath = currentInboundPath();
  return {
    method,
    path: pathWithQuery(url),
    request_id: requestId,
    ...(sessionId ? { session_id: sessionId } : {}),
    phase,
    relay_target: relayTarget,
    relay_origin: RELAY_ORIGIN,
    ...(originMethod ? { origin_method: originMethod } : {}),
    ...(originPath ? { origin_path: originPath } : {}),
    ...extra,
  };
}

/** Log an outbound HTTP call before it is sent (links relay to originating inbound request). */
export function logOutboundRequest(
  method: string,
  url: string,
  relayTarget: string,
  requestBody?: Record<string, unknown>,
  requestId?: string,
): void {
  const id = resolveRequestIdForLog(requestId);
  const originMethod = currentInboundMethod() ?? "";
  const originPath = currentInboundPath() ?? "";
  const fields = baseFields(method, url, relayTarget, id, "outbound_request", {
    headers: outboundRequestHeaders(id),
    ...(requestBody && Object.keys(requestBody).length > 0 ? { body: requestBody } : {}),
  });
  emit(
    "INFO",
    `${method} ${pathWithQuery(url)} outbound request request_id=${id} origin=${originMethod} ${originPath}`,
    fields,
  );
}

export function logOutboundResponse(
  method: string,
  url: string,
  status: number,
  ms: number,
  relayTarget: string,
  responseHeaders: Headers,
  responseBody: string | null,
  requestId?: string,
): void {
  const id = resolveRequestIdForLog(requestId);
  const sessionId = currentSessionId() ?? "";
  const ok = status >= 200 && status < 300;
  const fields = baseFields(method, url, relayTarget, id, "outbound_response", {
    status,
    ms,
    headers: filterResponseHeaders(responseHeaders),
    body: parseBody(responseBody),
    ...(ok ? {} : { error: resolveResponseError(status, responseBody) }),
  });
  const message = ok
    ? `${method} ${pathWithQuery(url)} ${status} outbound response request_id=${id} session_id=${sessionId}`
    : `${method} ${pathWithQuery(url)} ${status} outbound response request_id=${id} session_id=${sessionId} error=${resolveResponseError(status, responseBody) ?? ""}`;
  emit(ok ? "INFO" : "WARN", message, fields);
}

export function logOutboundFailure(
  method: string,
  url: string,
  ms: number,
  relayTarget: string,
  error: string,
  requestId?: string,
): void {
  const id = resolveRequestIdForLog(requestId);
  const sessionId = currentSessionId() ?? "";
  const fields = baseFields(method, url, relayTarget, id, "outbound_failed", {
    ms,
    error,
  });
  emit(
    "WARN",
    `${method} ${pathWithQuery(url)} outbound failed request_id=${id} session_id=${sessionId}`,
    fields,
  );
}

File: apps/python/src/exercises/web/outbound_http_logging.py
"""Structured outbound HTTP logging (http.client) for stack relays."""

from __future__ import annotations

import json
import logging
from typing import Any
from urllib.parse import urlparse

from exercises.web.correlation import merge_correlation

_LOG = logging.getLogger("http.client")
_RELAY_ORIGIN = "webserver-benchmark-python"
_HEADER_ALLOW = frozenset(
    {
        "x-request-id",
        "x-request-origin",
        "content-type",
        "content-length",
        "server",
        "date",
        "location",
    }
)


def _path_with_query(url: str) -> str:
    parsed = urlparse(url)
    pathname = parsed.path or "/"
    return f"{pathname}{parsed.query}" if parsed.query else pathname


def _parse_body(raw: str | None) -> dict[str, Any]:
    if not raw or not raw.strip():
        return {}
    text = raw.strip()
    try:
        if text.startswith("{"):
            parsed = json.loads(text)
            if isinstance(parsed, dict):
                return parsed
        if text.startswith("["):
            return {"_json": json.loads(text)}
    except json.JSONDecodeError:
        pass
    return {"_raw": text}


def resolve_response_error(status: int, response_body: str | None) -> str | None:
    if 200 <= status < 300:
        return None
    parsed = _parse_body(response_body)
    for key in ("error", "message"):
        value = parsed.get(key)
        if isinstance(value, str) and value.strip():
            return value.strip()
    raw = parsed.get("_raw")
    if isinstance(raw, str) and raw.strip():
        return raw.strip()
    return f"HTTP {status}"


def _filter_headers(headers: dict[str, str]) -> dict[str, str]:
    out: dict[str, str] = {}
    for key, value in headers.items():
        if key.lower() in _HEADER_ALLOW:
            out[key.lower()] = value
    return out


def _extra(**fields: Any) -> dict[str, Any]:
    extra = merge_correlation(fields)
    extra.setdefault("relay_origin", _RELAY_ORIGIN)
    return extra


def log_outbound_request(
    method: str,
    url: str,
    relay_target: str,
    *,
    request_body: dict[str, Any] | None = None,
    origin_method: str | None = None,
    origin_path: str | None = None,
) -> None:
    fields = _extra(
        method=method,
        path=_path_with_query(url),
        relay_target=relay_target,
        phase="outbound_request",
        origin_method=origin_method,
        origin_path=origin_path,
    )
    if request_body is not None:
        fields["body"] = request_body
    _LOG.info(
        "%s %s outbound request origin=%s %s",
        method,
        _path_with_query(url),
        origin_method or "",
        origin_path or "",
        extra=fields,
    )


def log_outbound_response(
    method: str,
    url: str,
    status: int,
    ms: int,
    relay_target: str,
    response_headers: dict[str, str],
    response_body: str | None,
    *,
    origin_method: str | None = None,
    origin_path: str | None = None,
) -> None:
    ok = 200 <= status < 300
    error = resolve_response_error(status, response_body)
    fields = _extra(
        method=method,
        path=_path_with_query(url),
        status=status,
        ms=ms,
        relay_target=relay_target,
        phase="outbound_response",
        origin_method=origin_method,
        origin_path=origin_path,
        response_headers=_filter_headers(response_headers),
        response_body=_parse_body(response_body),
    )
    if error:
        fields["error"] = error
    message = (
        f"{method} {_path_with_query(url)} {status} outbound response"
        if ok
        else f"{method} {_path_with_query(url)} {status} outbound response error={error or ''}"
    )
    if ok:
        _LOG.info(message, extra=fields)
    else:
        _LOG.warning(message, extra=fields)


def log_outbound_failure(
    method: str,
    url: str,
    ms: int,
    relay_target: str,
    error: str,
    *,
    origin_method: str | None = None,
    origin_path: str | None = None,
) -> None:
    _LOG.warning(
        "%s %s outbound failed",
        method,
        _path_with_query(url),
        extra=_extra(
            method=method,
            path=_path_with_query(url),
            ms=ms,
            relay_target=relay_target,
            phase="outbound_failed",
            error=error,
            origin_method=origin_method,
            origin_path=origin_path,
        ),
    )
